packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2590,12 +2590,12 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         return this.pendingStateManager.hasPendingMessages() || !this.emptyBatch;
     }
 
-    private updateDocumentDirtyState(dirty: boolean) {
+    private updateDocumentDirtyState(dirty: boolean, uploadingBlob: boolean = false) {
         if (this.attachState !== AttachState.Attached) {
             assert(dirty, 0x3d2 /* Non-attached container is dirty */);
         } else {
             // Other way is not true = see this.isContainerMessageDirtyable()
-            assert(!dirty || this.hasPendingMessages(),
+            assert(!dirty || this.hasPendingMessages() || uploadingBlob,
```

Comment:
I want to understand better thinking here and scoping.

I'm not sure we need that (design).
I'd go with a solution where dirty flag is fully controlled by ops.
Blob upload generates op and operation is not completed until op is acked.

You can argue that we mark document dirty when operation starts, but this is really undefined - how does one define "start"? There is no requirement anywhere in the system that when first part (synchronous part of the call) is over, it's a "start". And given that (at least current) design does not apply this change immediately to date mode, data model is actually never dirty through the whole process. It's atomically switched to "image is there" when op is acked.

Maybe I do not fully understand concerns, but they feel like similar to "I have something in the app happening in response to user action, data model is not yet dirty, but I want it dirty". It does feel like it's not really runtime problem.

Maybe it's only for offline mode? If so, I really want to call out that it's not ideal to have two different models RE how changes are applied to data model (synchronous & async). I'd want to have a discussion - can we have same flow and not complicate designs for one side of implementation?

If we had fully synchronous flow for image to apply to local data mode, then we would likely need something like that, but (as comment below points out) you need to work with counters, not booleans, similar to how this.hasPendingMessages() works - it looks a ALL pending ops. All of this code dealing with blobs should look at ALL pending blob uploads.

Reply:
Does changing solution to check for BlobManager.pendingBlobs array solve your concerns about sync vs async flow? I'm still not completely sure that we should only check isDirty state during containerRuntime constructor though

Reply:
The host needs to know about inflight blob operations, otherwise it can lead to dataloss. The mechanism we have for the host to determine if there are inflight operation on a document is save/dirty

Reply:
@dannimad - yes, that would be the better place, but I have a lot concerns here (see below long comment with a bunch of replies). Moving it to blob manage is just first step. We should re-evaluate all transition and make sure that system has some consistency. This includes small things (like treating blobs uploaded in "attaching" and "attached" states the same), but also big things (if it's dirty, then set of local changes is non-empty).